src/connector/src/sink/redis.rs

```diff
@@ -12,39 +12,302 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::collections::HashMap;
+use std::fmt::{Debug, Formatter};
+
 use async_trait::async_trait;
-use risingwave_common::array::StreamChunk;
+use itertools::Itertools;
+use redis::{ConnectionLike, RedisResult, Value};
+use risingwave_common::array::{Op, RowRef, StreamChunk};
 use risingwave_common::catalog::Schema;
+use risingwave_common::types::to_text::ToText;
+
+use crate::sink::{Result, Sink, SinkError};
 
-use crate::sink::{Result, Sink};
+pub const REDIS_SINK: &str = "redis";
 
 #[derive(Clone, Debug)]
-pub struct RedisConfig;
+pub struct RedisConfig {
+    pub endpoint: String,
+}
 
-#[derive(Debug)]
-pub struct RedisSink;
+impl RedisConfig {
+    pub fn from_hashmap(map: HashMap<String, String>) -> Result<Self> {
+        let endpoint = map
+            .get("endpoint")
+            .ok_or_else(|| SinkError::Config("endpoint".to_string()))?
+            .to_string();
+        Ok(RedisConfig { endpoint })
+    }
+}
+
+pub struct RedisSink {
+    config: RedisConfig,
+    schema: Schema,
+    // connection to redis, one per executor
+    client: Option<redis::Client>,
+    conn: RedisConnectionImpl,
+    // the command pipeline for write-commit
+    pipe: redis::Pipeline,
+    batch_id: u64,
+    epoch: u64,
+    update_cache: Vec<String>,
+    pk_indices: Vec<usize>,
+}
+
+pub enum RedisConnectionImpl {
+    Redis(redis::Connection),
+    Mock(MockConnection),
+}
+
+impl Debug for RedisSink {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("RedisSink")
+            .field("config", &self.config)
+            .field("schema", &self.schema)
+            .field("client", &self.client)
+            .field("batch_id", &self.batch_id)
+            .field("epoch", &self.epoch)
+            .field("update_cache", &self.update_cache)
+            .field("pk_indices", &self.pk_indices)
+            .finish()
+    }
+}
 
 impl RedisSink {
-    pub fn new(_cfg: RedisConfig, _schema: Schema) -> Result<Self> {
-        todo!()
+    pub fn new(cfg: RedisConfig, schema: Schema, pk_indices: Vec<usize>) -> Result<Self> {
+        let client = redis::Client::open(cfg.endpoint.clone())?;
+        let conn = client.get_connection()?;
+        let pipe = redis::pipe();
+        Ok(Self {
+            config: cfg,
+            schema,
+            client: Some(client),
+            conn: RedisConnectionImpl::Redis(conn),
+            pipe,
+            batch_id: 0,
+            epoch: 0,
+            update_cache: Vec::new(),
```

Comment:
update_cache can only take 1 element for validation, better `Vec::with_capcity(1)`